From 3f3bc52661776239e822ec8cd28ad9bf309874e2 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 18 Aug 2018 22:03:33 -0400
Subject: [PATCH] SpawnEggMeta#setSpawnedEntity API

lets you copy an entities data into a spawn egg.
Partial data is supported through a predicate, letting MC
follow normal spawn behavior in the summon phase.
---
 .../inventory/CraftMetaSpawnEgg.java          | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 815cca538..6dfbc5f6f 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -2,6 +2,7 @@ package org.bukkit.craftbukkit.inventory;
 
 import com.google.common.collect.ImmutableMap.Builder;
 import java.util.Map;
+
 import net.minecraft.server.MinecraftKey;
 import net.minecraft.server.NBTBase;
 import net.minecraft.server.NBTTagCompound;
@@ -229,6 +230,24 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
         return spawnedType != null;
     }
 
+    // Paper start
+    private static final String[] removeKeys = {
+        "UUIDLeast", "UUIDMost", "Pos", "PortalCooldown", "Paper.Origin", "Paper.FromMobSpawner", "Passengers", "Dimension"
+    };
+
+    @Override
+    public void setSpawnedEntity(org.bukkit.entity.Entity entity, java.util.function.Predicate<String> keyFilter) {
+        entityTag = ((org.bukkit.craftbukkit.entity.CraftEntity) entity).getHandle().save(new NBTTagCompound());
+        for (String removeKey : removeKeys) {
+            entityTag.remove(removeKey);
+        }
+        if (keyFilter != null) {
+            entityTag.map.keySet().removeIf(keyFilter);
+        }
+        entityTag.setBoolean("Paper.CustomSpawnEgg", true);
+    }
+    // Paper end
+
     @Override
     public EntityType getSpawnedType() {
         throw new UnsupportedOperationException("Must check item type to get spawned type");
-- 
2.24.1

